### PR TITLE
[SPARK-27269][SQL] File source v2 should validate data schema only

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?

Currently, File source v2 allows each data source to specify the supported data types by implementing the method `supportsDataType` in `FileScan` and `FileWriteBuilder`.

However, in the read path, the validation checks all the data types in `readSchema`, which might contain partition columns. This is actually a regression. E.g. Text data source only supports String data type, while the partition columns can still contain Integer type since partition columns are processed by Spark.

This PR is to:
1. Refactor schema validation and check data schema only.
2. Filter the partition columns in data schema if user specified schema provided.

## How was this patch tested?

 Unit test

### CODE REVIEW

**Comment by gengliangwang** [comment]
@cloud-fan @dongjoon-hyun @HyukjinKwon

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileTable.scala [hunk 1]
```diff
@@ -46,17 +46,27 @@ abstract class FileTable(
...
+    }.asNullable
 
   override lazy val schema: StructType = {
     val caseSensitive = sparkSession.sessionState.conf.caseSensitiveAnalysis
     SchemaUtils.checkColumnNameDuplication(dataSchema.fieldNames,
       "in the data schema", caseSensitive)
+    dataSchema.foreach { field =>
+      if (!supportsDataType(field.dataType)) {
+        throw new AnalysisException(
+          s"$name data source does not support ${field.dataType.catalogString} data type.")
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileTable.scala]: Here, `name` is `Table.name`, isn't? Previously, this was `formatName`.

**Review by HyukjinKwon — APPROVED** [review]
Looks good except one Dongjoon's comment and the test failure.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileTable.scala [hunk 2]
```diff
@@ -46,17 +46,27 @@ abstract class FileTable(
...
 
-  lazy val dataSchema: StructType = userSpecifiedSchema.orElse {
-    inferSchema(fileIndex.allFiles())
-  }.getOrElse {
-    throw new AnalysisException(
-      s"Unable to infer schema for $name. It must be specified manually.")
-  }.asNullable
+  lazy val dataSchema: StructType = userSpecifiedSchema.map { schema =>
+      val partitionSchema = fileIndex.partitionSchema
+      val equality = sparkSession.sessionState.conf.resolver
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileTable.scala]: `equality` -> `resolver`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala [hunk 3]
```diff
@@ -37,22 +37,6 @@ abstract class FileScan(
...
-  /**
-   * The string that represents the format that this data source provider uses. This is
-   * overridden by children to provide a nice alias for the data source. For example:
-   *
-   * {{{
-   *   override def formatName(): String = "ORC"
-   * }}}
-   */
-  def formatName: String
-
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala]: @gengliangwang . This looks irrelevant to this PR. And, I cannot see the explanation about moving `formatName`. For me, `CSVScan` naturally has `def formatName`. Why do we need to move `formatName`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileTable.scala (hunk 2 above)
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileTable.scala]: The naming is following DataSource.scala line 185. I think it is OK.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala (hunk 3 above)
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala]: The `formatName` was added for showing the Exception message.
https://github.com/apache/spark/pull/23714#pullrequestreview-203520602

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala]: Thanks. Yes. It seems to be moved together correctly.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileTable.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileTable.scala]: There is only [one line](https://github.com/apache/spark/blame/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala#L111) written two year ago. All the other new instances use `resolver = sparkSession.sessionState.conf.resolver` (more than 7).

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileTable.scala]: If you search with `conf.resolver`, there are more instances.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileTable.scala [hunk 4]
```diff
@@ -46,17 +46,27 @@ abstract class FileTable(
...
+      StructType(schema.filterNot(f => partitionSchema.exists(p => equality(p.name, f.name))))
+    }.orElse {
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileTable.scala]: Indentation? (https://github.com/databricks/scala-style-guide#pattern-matching)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala [hunk 5]
```diff
@@ -76,13 +60,5 @@ abstract class FileScan(
...
 
-  override def toBatch: Batch = {
-    readSchema.foreach { field =>
-      if (!supportsDataType(field.dataType)) {
-        throw new AnalysisException(
-          s"$formatName data source does not support ${field.dataType.catalogString} data type.")
-      }
-    }
-    this
-  }
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala]: Shall we update the `import` according to this deletions?
```scala
-import org.apache.spark.sql.{AnalysisException, SparkSession}
+import org.apache.spark.sql.SparkSession
 import org.apache.spark.sql.execution.PartitionedFileUtil
 import org.apache.spark.sql.execution.datasources._
 import org.apache.spark.sql.sources.v2.reader.{Batch, InputPartition, Scan}
-import org.apache.spark.sql.types.{DataType, StructType}
+import org.apache.spark.sql.types.StructType
 import org.apache.spark.sql.util.CaseInsensitiveStringMap
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVDataSourceV2.scala [hunk 6]
```diff
@@ -41,13 +41,3 @@ class CSVDataSourceV2 extends FileDataSourceV2 {
...
 }
-
-object CSVDataSourceV2 {
-  def supportsDataType(dataType: DataType): Boolean = dataType match {
-    case _: AtomicType => true
-
-    case udt: UserDefinedType[_] => supportsDataType(udt.sqlType)
-
-    case _ => false
-  }
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVDataSourceV2.scala]:
```
-import org.apache.spark.sql.types._
+import org.apache.spark.sql.types.StructType
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcDataSourceV2.scala [hunk 7]
```diff
@@ -42,19 +42,3 @@ class OrcDataSourceV2 extends FileDataSourceV2 {
...
-    case ArrayType(elementType, _) => supportsDataType(elementType)
-
-    case MapType(keyType, valueType, _) =>
-      supportsDataType(keyType) && supportsDataType(valueType)
-
-    case udt: UserDefinedType[_] => supportsDataType(udt.sqlType)
-
-    case _ => false
-  }
-}
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcDataSourceV2.scala]: ditto.
```
-import org.apache.spark.sql.types._
+import org.apache.spark.sql.types.StructType
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcScan.scala [hunk 8]
```diff
@@ -43,10 +43,4 @@ case class OrcScan(
     OrcPartitionReaderFactory(sparkSession.sessionState.conf, broadcastedConf,
       dataSchema, fileIndex.partitionSchema, readSchema)
   }
-
-  override def supportsDataType(dataType: DataType): Boolean = {
-    OrcDataSourceV2.supportsDataType(dataType)
-  }
-
-  override def formatName: String = "ORC"
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcScan.scala]:
```
-import org.apache.spark.sql.types.{DataType, StructType}
+import org.apache.spark.sql.types.StructType
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/FileTableSuite.scala [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+      // However, the partition schema is handled by Spark, so it is allowed to contain
+      // Integer data type here.
+      val table = new DummyFileTable(spark, options, Seq(pathName), expectedDataSchema, None)
+      assert(table.dataSchema == expectedDataSchema)
+      val expectedPartitionSchema = StructType(Seq(StructField("p", IntegerType, true)))
+      assert(table.fileIndex.partitionSchema ==  expectedPartitionSchema)
+    }
+  }
+
+  test("Returns correct data schema when user specified schema contains partition schema") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/FileTableSuite.scala]: Thank you for adding this test.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/FileTableSuite.scala [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+      df.write.partitionBy("p").text(pathName)
+      val options = new CaseInsensitiveStringMap(Map("path" -> pathName).asJava)
+      val expectedDataSchema = StructType(Seq(StructField("v", StringType, true)))
+      // DummyFileTable doesn't support Integer data type.
+      // However, the partition schema is handled by Spark, so it is allowed to contain
+      // Integer data type here.
+      val table = new DummyFileTable(spark, options, Seq(pathName), expectedDataSchema, None)
+      assert(table.dataSchema == expectedDataSchema)
+      val expectedPartitionSchema = StructType(Seq(StructField("p", IntegerType, true)))
+      assert(table.fileIndex.partitionSchema ==  expectedPartitionSchema)
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/FileTableSuite.scala]: nit. additional space after `==`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileTable.scala [hunk 11]
```diff
@@ -72,6 +82,22 @@ abstract class FileTable(
    * Spark will require that user specify the schema manually.
    */
   def inferSchema(files: Seq[FileStatus]): Option[StructType]
+
+  /**
+   * Returns whether this format supports the given [[DataType]] in write path.
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileTable.scala]: `write` -> `read/write`.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
@gengliangwang . Sorry for missing the regression last time. 
This PR seems to have a different issue. Previously, we check the schema in read-path via `toBatch`. For now, it's removed. Currently, we only remove partition columns by names and we don't check the column types in the data schema.

**Comment by HyukjinKwon** [comment]
@dongjoon-hyun, do you mean the issue is that we don't check the schema in `toBatch` but lazily within `Table`? The data schema looks being checked via overridden `schema` at `FileTable` in read-path.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileTable.scala [hunk 12]
```diff
@@ -57,6 +61,12 @@ abstract class FileTable(
     val caseSensitive = sparkSession.sessionState.conf.caseSensitiveAnalysis
     SchemaUtils.checkColumnNameDuplication(dataSchema.fieldNames,
       "in the data schema", caseSensitive)
+    dataSchema.foreach { field =>
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileTable.scala]: @dongjoon-hyun The schema validation of read path is moved here.

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileTable.scala]: My bad. When I tested in `FileTableSuite` yesterday, I missed this part. Ya. Thanks, @gengliangwang and @HyukjinKwon .

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileTable.scala [hunk 13]
```diff
@@ -46,7 +46,11 @@ abstract class FileTable(
       sparkSession, rootPathsSpecified, caseSensitiveMap, userSpecifiedSchema, fileStatusCache)
   }
 
-  lazy val dataSchema: StructType = userSpecifiedSchema.orElse {
+  lazy val dataSchema: StructType = userSpecifiedSchema.map { schema =>
+    val partitionSchema = fileIndex.partitionSchema
+    val equality = sparkSession.sessionState.conf.resolver
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileTable.scala]: This seems to be the last comment which is unresolved (https://github.com/apache/spark/pull/24203#discussion_r268951495)
We had better make the code as consistent as possible.

**Review by dongjoon-hyun — APPROVED** [review]
Only one nit comment about naming. Thank you, @gengliangwang .

**Comment by HyukjinKwon** [comment]
Merged to master.
